### ISSUE TITLE
Retiring Canned Examples?

### ISSUE THREAD

**Issue by pzinn** [Macaulay2/M2#1924]
I don't quite know what the history of Canned Examples is, but by now, we have two ways of generating examples: either by outputtting them to a .out file, or by using `capture`. I think that should be enough to cover all cases. I philosophically dislike Canned Examples because they're hard-coded and because they're not  mode-agnostic. About the latter point, what I mean is
(1) they look ugly in WebApp mode and
(2) they force us (well, me, really) to write ugly hacks in the handling of example output since `capture` produces examples in the current output mode (which I think is a good thing) whereas CannedExamples are always in standard mode. I will have to submit a PR anyway because https://github.com/Macaulay2/M2/commit/90960b09e1e841f8c643a91f1310b36edac90d46#commitcomment-46835042 broke `capture` in WebApp mode but I don't look forward to having to treat CannedExamples as well; originally when I rewrote `separateM2output` (the function that parses example output) I tried to make it as clean as possible and now it's kind of sloppy again.
So in conclusion I'm ready to go through the Core and check all CannedExamples and see if there's some way to replace them with normal examples. Suggestions of alternatives welcome.

**Comment by d-torrance** [comment]
I think the main purpose of canned examples is for situations when generating the example results in an error, and consequently a build failure.  I can think of at least three reasons why this might happen:

* It makes sense to actually include an error in the example to actually document the feature we're documenting, e.g., `assert`:

```m2
     assert x prints an error message if x isn't true.

     +--------------------------------------------------+
     |  i1 : assert( (2+2) === 4)                       |
     +--------------------------------------------------+
     |  i2 : assert(rank matrix {{1, 2}, {2, 4}} == 2)  |
     |  stdio:2:1:(3): error: assertion failed          |
     +--------------------------------------------------+
```

* The example depends on some program that isn't a Macaulay2 build dependency, e.g., `(html, Digraph)`:

```m2
    Uses graphviz to create an .svg representation of G, which is returned as
     a string.

     +-------------------------------------------------------------------------------------------------------------------------+
     |  i2 : html completeGraph 2                                                                                              |
     |                                                                                                                         |
     |  o2 = <?xml version="1.0" encoding="UTF-8" standalone="no"?>                                                            |
     |       <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"                                                                    |
     |        "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">                                                              |
     |       <!-- Generated by graphviz version 2.43.0 (0)                                                                     |
     |       <!-- Title: G Pages: 1 -->                                                                                        |
     |       <svg width="62pt" height="116pt"                                                                                  |
     |        viewBox="0.00 0.00 62.00 116.00" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">  |
     |       <g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 112)">                                   |
     |       <title>G</title>                                                                                                  |
     |       <polygon fill="white" stroke="transparent" points="-4,4 -4,-112 58,-112 58,4 -4,4"/>                              |
     |       <!-- 0 -->                                                                                                        |
     |       <g id="node1" class="node">                                                                                       |
     |       <title>0</title>                                                                                                  |
     |       <ellipse fill="none" stroke="black" cx="27" cy="-90" rx="27" ry="18"/>                                            |
     |       <text text-anchor="middle" x="27" y="-86.3" font-family="Times,serif" font-size="14.00">0</text>                  |
     |       </g>                                                                                                              |
     |       <!-- 1 -->                                                                                                        |
     |       <g id="node2" class="node">                                                                                       |
     |       <title>1</title>                                                                                                  |
     |       <ellipse fill="none" stroke="black" cx="27" cy="-18" rx="27" ry="18"/>                                            |
     |       <text text-anchor="middle" x="27" y="-14.3" font-family="Times,serif" font-size="14.00">1</text>                  |
     |       </g>                                                                                                              |
     |       <!-- 0&#45;&#45;1 -->                                                                                             |
     |       <g id="edge1" class="edge">                                                                                       |
     |       <title>0&#45;&#45;1</title>                                                                                       |
     |       <path fill="none" stroke="black" d="M27,-71.7C27,-60.85 27,-46.92 27,-36.1"/>                                     |
     |       </g>                                                                                                              |
     |       </g>                                                                                                              |
     |       </svg>                                                                                                            |
     +-------------------------------------------------------------------------------------------------------------------------+
```
* Dealing with examples that are known to fail regularly on various architectures.  This is a downstream issue I deal with in the Debian package -- see https://salsa.debian.org/science-team/macaulay2/-/blob/debian/master/debian/patches/use-canned-examples.patch.  If/when we switch to an architecture-independent database library, then this will become a non-issue as we can just generate all the documentation on amd64 and use that for the other architectures.

I don't like canned examples, either, but they seem like a necessary evil to avoid build failures.

I suppose .out files would accomplish the same thing, but they seem less granular.  Is it possible to mix-and-match them, i.e., some canned and some generated files for the same key?  And getting them in the first place is kind of a hassle -- we can set `CacheExampleOutput` for the entire package and then delete the ones we don't need, or copy just the ones we need from the installation directory to the source directory after build.  Neither seem very straightforward, unless I'm missing something.

**Comment by mahrud** [comment]
> I will have to submit a PR anyway because [90960b0#commitcomment-46835042](https://github.com/Macaulay2/M2/commit/90960b09e1e841f8c643a91f1310b36edac90d46#commitcomment-46835042) broke `capture` in WebApp mode

By broken all you mean is that it doesn't drop the first and last element, correct? Have you tried modifying your regex instead? You can have it separate either at `webAppCellTag` or at the regular boundary. Didn't #1553 help with this?

> originally when I rewrote `separateM2output` (the function that parses example output) I tried to make it as clean as possible and now it's kind of sloppy again.

We shouldn't always drop the first and last element after separating the output. That would be sloppy.

**Comment by DanGrayson** [comment]
How would I take a look at an ugly canned example in your web app?

**Comment by pzinn** [comment]
It might take a while before I have time to work on this issue. In any case I'm not going to rush to make any changes. For now two comments:
- Concerning examples that produce errors, isn't it possible to have a flag, something like `-- ignore-error`? after all we already have a no-cpature flag...
- Examples that require external programs: I'm a bit confused about this. After all, there are whole packages that depend on such external programs (e.g. Maple), and theirs examples are not canned and everything is fine. So why do we need to can some?

**Comment by DanGrayson** [comment]
I don't much like the idea of canning examples, but for external programs that we don't include or require to be installed, it is a good idea to cache the example output.

Formerly, if we required anything external, we'd have to be prepared to download it and compile it, as part of the build process 
 ---  sometimes that wasn't possible, because of restrictive licensing.  Now there are probably more and more external packages that we can depend on finding on all the systems where we have users, but we have users on a broad array of system types, and we should take care not to shut out users on some of those systems.

**Comment by mahrud** [comment]
> * Concerning examples that produce errors, isn't it possible to have a flag, something like `-- ignore-error`? after all we already have a no-capture flag...

I added `no-capture-flag` as a temporary hack until capture can handle more cases and I'd like to avoid propagating similar hacks.

> * Examples that require external programs: I'm a bit confused about this. After all, there are whole packages that depend on such external programs (e.g. Maple), and theirs examples are not canned and everything is fine. So why do we need to can some?

This is not true, they are also canned, just in a different way:
https://github.com/Macaulay2/M2/tree/master/M2/Macaulay2/packages/Bertini/examples

Fundamentally, I don't understand why a canned example is any different. Could you link an ugly example?

**Comment by pzinn** [comment]
> > * Examples that require external programs: I'm a bit confused about this. After all, there are whole packages that depend on such external programs (e.g. Maple), and theirs examples are not canned and everything is fine. So why do we need to can some?
> 
> This is not true, they are also canned, just in a different way:
> https://github.com/Macaulay2/M2/tree/master/M2/Macaulay2/packages/Bertini/examples
> 
They are *not* `CannedExamples`s. I'm perfectly fine with having generated examples be cached, of course.

**Comment by pzinn** [comment]
> Fundamentally, I don't understand why a canned example is any different. Could you link an ugly example?
I don't have time to look for a good example, but just compare `help assert` and `help det` on the web interface.

**Comment by DanGrayson** [comment]
> * Concerning examples that produce errors, isn't it possible to have a flag, something like `-- ignore-error`? after all we already have a no-capture flag...

What I do in the documentation for "the debugger" is to run `stopIfError = false`.  That will put the error messages into the example output.

**Comment by d-torrance** [comment]
> What I do in the documentation for "the debugger" is to run `stopIfError = false`. That will put the error messages into the example output.

Ah, I didn't know that was a thing!  That makes my first point moot.  I'll submit a PR adding this to the `assert` docs so it can use generated examples.

**Comment by DanGrayson** [comment]
Should probably turn on not capturing, too, as `stopIfError` is a global variable.

**Comment by d-torrance** [comment]
> Should probably turn on not capturing, too, as `stopIfError` is a global variable.

That's already taken into account in `isCapturable`:

https://github.com/Macaulay2/M2/blob/deb37e0a86df12c65ea1cdb7a83d960e7c1c06f7/M2/Macaulay2/m2/examples.m2#L122
